### PR TITLE
codemodedev's GetGPSCoordinates Spice

### PR DESCRIPTION
**Pull request by codemodedev**
**What does your Instant Answer do?**
Displays the GPS coordinates for any city, and a link to those coordinates in OpenStreetMap.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
It's quicker and easier than wading through search results that may or may not actually show the specific coordinates for the requested city.

**What is the data source for your Instant Answer? (Provide a link if possible)**
Wikipedia

**Why did you choose this data source?**
It appears to have GPS coordinates for every major city and many small towns as well, with a fast open API

**Are there any other alternative (better) data sources?**
Possibly, but none that I found had an API or at least not one as fast as Wikipedia's.

**What are some example queries that trigger this Instant Answer?**
gps for denver, chicago gps coordinates, gps coordinates for paoli pa

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
cartographers, people who work with weather stations, data geeks

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
no

**Which existing Instant Answers will this one supersede/overlap with?**
I don't see any (at least in the Spice tree?) that mention either GPS or coordinates

**Are you having any problems? Do you need our help with anything?**
When running the tests, some of my negative tests are failing.  For example, "gps for denver" is valid for my IA, but "gps for sale" is not.  This works properly in the browser (the IA is simply not displayed), but in the test it fails, because my code intentionally returns Spice.failed() from the JS in cases where the API returns no GPS coordinates, but the command-line test can't test for such JS returns (according to the DDG guys in the Gitter chat).
The test output is:

  Failed test 'Checking for not matching on gps for sale'
  at /home/codio/.../Test/Block.pm line 62.
  got: 'DDG::ZeroClickInfo::Spice=HASH(0x365ccd8)'
  expected: undef

For now I've just commented out those types of negative tests (other negative tests work properly, such as 'gps coordinate conversion' => undef).  I guess that's fine, but should there be a way to test for them that uses the JS output in the test?

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
No API key or costs are involved

**Where did you hear about DuckDuckHack? (For first time contributors)**
Just browsing around the DDG site and the about-us page

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**

## Checklist
Please place an 'X' where appropriate.

```
[x] Added metadata and attribution information
[x] Wrote test file and added to t/ directory
[x] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[x] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[x] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Linux (LinuxMint 17)
        [x] Google Chrome
        [x] Firefox

    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [x] Google Chrome
        [x] Firefox
        [x] Opera
        [] IE 8
        [x] IE 9
        [x] IE 10
        [x] IE 11

    - Windows XP
        [x] IE 8

    - Mac OSX
        [x] Google Chrome
        [x] Firefox
        [x] Opera
        [x] Safari

    - iOS (iPhone)
        [x] Safari Mobile
        [x] Google Chrome
        [x] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [] Google Chrome
        [] Opera
```
screenshot

### CODE REVIEW

**Comment by MrChrisW** [comment]
@codemodedev Thanks for your contribution! We'll be sure to review this shortly.

File: lib/DDG/Spice/GetGPSCoordinates.pm [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    s!(,|\s+)TN$!$1Tennessee!i;
+    s!(,|\s+)TX$!$1Texas!i;
+    s!(,|\s+)UT$!$1Utah!i;
+    s!(,|\s+)VT$!$1Vermont!i;
+    s!(,|\s+)VI$!$1Virgin Islands!i;
+    s!(,|\s+)VA$!$1Virginia!i;
+    s!(,|\s+)WA$!$1Washington!i;
+    s!(,|\s+)WV$!$1West Virginia!i;
+    s!(,|\s+)WI$!$1Wisconsin!i;
+    s!(,|\s+)WY$!$1Wyoming!i;
```
Comment (MrChrisW) [inline lib/DDG/Spice/GetGPSCoordinates.pm]: @codemodedev You may be able to use the [Locale-SubCountry](https://github.com/cebjyre/Locale-SubCountry) package for these abbreviations. See [SalesTax](https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/lib/DDG/Spice/SalesTax.pm) as an example.

File: t/GetGPSCoordinates.t [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        caller => 'DDG::Spice::GetGPSCoordinates'
+    ),
+    'denver gps coordinates' => test_spice(
+        '/js/spice/get_gpscoordinates/Denver',
+        call_type => 'include',
+        caller => 'DDG::Spice::GetGPSCoordinates'
+    ),
+    # Try to include some examples of queries on which it might
+    # appear that your answer will trigger, but does not.
+    #'gps for sale' => undef,
```
Comment (MrChrisW) [inline t/GetGPSCoordinates.t]: Comments can be removed.

File: share/spice/get_gpscoordinates/get_gpscoordinates.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
+(function (env) {
+    "use strict";
+    env.ddg_spice_get_gpscoordinates = function(api_result){
+
+        if (api_result.error) {
```
Comment (MrChrisW) [inline share/spice/get_gpscoordinates/get_gpscoordinates.js]: Also need to check if we get a vaild response at all eg.  `!api_result | api_result.error`

File: lib/DDG/Spice/GetGPSCoordinates.pm [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+# Metadata.  See https://duck.co/duckduckhack/metadata for help in filling out this section.
+name "GetGPSCoordinates";
+source "Wikipedia";
+#icon_url "";
+description "Displays the GPS coordinates for the requested city, with a link to OpenStreetMap.";
+primary_example_queries "gps for paoli, pa", "gps coordinates for philadelphia";
+secondary_example_queries "denver gps coordinates";
+# Uncomment and complete: https://duck.co/duckduckhack/metadata#category
+category "geography";
+# Uncomment and complete: https://duck.co/duckduckhack/metadata#topics
```
Comment (MrChrisW) [inline lib/DDG/Spice/GetGPSCoordinates.pm]: Remove comments above, leave only ABSTRACT

**Comment by MrChrisW** [comment]
@codemodedev This looks ok, however it would be preferable to use an API that returns formatted JSON with `lat` and `long` properties, this would avoid doing all the regex searching/parsing of the Wikipedia response. 

A good free (hit limit) API is the GeoNames database: 

Example: http://api.geonames.org/searchJSON?formatted=true&q=Pennsylvania&maxRows=10&lang=es&username=demo&style=full

Docs: http://www.geonames.org/export/web-services.html

**Comment by moollaza** [comment]
@codemodedev thanks for contributing. This is great, but I think I mentioned before that we should probably just have our existing Maps IA handle these queries, and then update the frontend to show the lat/lon for the map pins.

As @MrChrisW said parsing Wiki with regex's really isn't ideal and probably won't have as good coverage/relevancy checking as our maps data.

@bsstoner what are your thoughts on this? Should we implements this is a separate IA (it looks like we could copy Maps.pm?) with diffferent triggers and set it up to show the Lat/Lon, or should this just be an improvement to our existing Maps IA?

**Comment by moollaza** [comment]
@codemodedev we've made some recent changes to our Maps and sources and we now support these kinds of queries and results, but its slightly broken it seems right now and needs some repairing. Would you be interested in adding some triggers. I'm not sure if the open source JS is where the existing bug lies, but you're welcome to take a look. Here's an example of the functionality: 

https://duckduckgo.com/?q=longitude+of+toronto+canada&t=ffab&ia=maps

Note the `NaN` value in the map pin, which is a bug!

**Pull request closed without merging**